### ISSUE TITLE
Re-visit how new ontologies can be added to Metacat's annotation subprocessor

### ISSUE THREAD

**Issue by amoeba** [NCEAS/metacat#1498]
[Note: Whatever is done here needs to be applied to the `d1_cn_index_process` too]

We're starting to create new one-off ontologies and even have other groups use EML semantic annotations and both cases are creating annotations using ontologies that aren't in the pre-defined list we index with. The annotations are still indexed but we won't perform query expansion to materialize their parent classes because the terms aren't in the list.

We need a way that we and other Metacat admins can relatively simply added ontologies to Metacat. And we might think about whether we could relax the restriction I put in place and allow query expansion on arbitrary ontologies by fetching them on-demand. The latter may be the route to go because (1) groups will use Metacat & EML Annotations in ways we don't know about and won't know they aren't getting the full benefit and (2) even for us, having more manual steps in the way of adding new ontologies to Metacat slows us down.

Some possible routes to go down include:

- Load new ontologies from the filesystem (so long as this works). This'd mean editing the spring context file + adding an OWL file
- Build out an admin UI that lets the user do something like the above
- Load ontologies on demand as terms come in. This is how the old Annotator code worked and I disabled it since I didn't see a big use case and thought it could create a bad/insecure experience

**Comment by mbjones** [comment]
Yeah, we need to figure out an approach here. I agree that loading arbitrary ontologies for query expansion has both security risks and logical reasoning entainlments that should be carefully considered. Right now, we carefully curate the metadata standards that we support by registering them in the object formats list, storing a definitive copy in our Metacat release, and registering those in our Metacat xml-catalog configuration so the local copies are used in preference to network copies. A similar approach could be applied to ontologies, where they are registered and cached (as ttl files?) so we have canonical versions, which allows us to review them for safety and reasoning repercussions before they are used in indexing. or maybe this is unnecessary? Can we trust the bioportal copy registered with a given namespace?

Whenever a change to the recognized list of ontologies occurs, we likely need to reindex all content with annotations that relate to the new statements that are added. So, we probably also need to consider how to trigger this without overloading the indexer.

**Comment by amoeba** [comment]
I'm going to spend some time this week working on this.

My preferred approach here would be to make changing the set of ontologies as easy as changing the code, re-building, and re-deploying. But since full releases take many months, we really need to be able to hot-patch the filesystem (update context file, add/remove OWL files). Even if this cycle were faster, it'd imply downtime which is also not great. We need something that would affect the running system quickly (a few minutes) and without downtime.

How annotation indexing works right now is that EML annotations get passed over to an OntologyModelService singleton that's created at startup. The OntologyModelService owns a Jena OntologyModel which gets pre-loaded with a list of ontologies defined in its Spring context file.

Some constraints we have are:

- The change needs to be applicable to both `metacat-index` and `d1_cn_index_processor`. ie we can't use /var/metacat
- Updating the ontologies list really shouldn't require downtime of Metacat/REST API. It's probably okay if index processing is paused for a moment

Some ideas for a path forward:

1. Re-load every ontology on every index task rather than just at startup. This might be way too slow but maybe it's not and I prematurely optimized?
2. Look into how/if Spring can help us reload things when the filesystem changes or if we can easily implement this ourselves.
  a. If it's not easy to get Spring to reload things, we could establish a sentinel file in the WAR deploy directory called something like `.reload`. The OntologyModelService would check for this on every index job and then refresh and clean up the sentinel. The sentinel could also be kept in in the database, hazelcast, or anywhere really. Maybe there's already some convention out their in the Java/Spring world
3. Set up a Metacat admin API or D1 REST API mechanism to force a reload, similar to how we rebuild the stylesheet cache. I don't think either of those APIs can talk _directly_ to `metacat-index` or `d1_cn_index_processor` but maybe there's a way.

**Comment by mbjones** [comment]
@amoeba looking forward to seeing the plan for this. Keep in mind that we are also looking at refactoring the indexing system into a horizontally scalable system that allows many index workers to operate in parallel from a task queue. In that situation, would you imagine the singleton is at the worker level, rather than in metacat-index? I think each worker would need a copy of the Jena OntologyModel and they would all need to update when a new ontology is added or changed.

**Comment by amoeba** [comment]
Right, good point to bring up. I'd imagine each worker would keep its own singleton. A distributed indexing system simplifies my above list because we'd just need to kill existing workers and deploy new ones to see new behavior. It'd let us make quick changes without downtime.

**Comment by mbjones** [comment]
Yeah, that sounds good. Another item that needs to be considered is how an ontology might require existing documents to be reindexed, and so we may need an ontology update to schedule new index tasks once the workers are updated. These could be done in the background at lower priority than index jobs coming directly from users (e.g., via MetacatUI submissions).
